### ISSUE TITLE
Behavior of pass in MonadWriter fails to match WriterT

### ISSUE THREAD

**Issue by paf31** [fpco/monad-unlift#3]
I'm not sure if the goal of this library is to match the semantics of the `transformers` instances, but `pass` definitely does something different in `WriterRefT`. My understanding is that `pass` should only apply the returned function to those results logged in the scope of the action passed as its argument, but `modifyRef'` seems to apply it to all results, including those logged _before_ the action was run.

I noticed this when trying to use `WriterRefT` as a low-overhead replacement for `WriterT` in `purescript`: 

https://github.com/purescript/purescript/issues/1297#issuecomment-151255131

This also affects `censor`, which I think is implemented in terms of `pass`.

**Comment by paf31** [comment]
I fixed this in `purescript` by using a fresh `IORef` inside `listen` and `pass`. I expect the same would work here.

See https://github.com/purescript/purescript/pull/1572/files

**Comment by snoyberg** [comment]
Yes, the goal is to match semantics, and if it doesn't, it's a bug. Would you be able to send a PR?

**Comment by paf31** [comment]
I don't really use this library, I just spotted this when I used it to prototype the replacement of `WriterT`, but I can have a go if I end up getting to it first.
